pyteal/ast/abi/uint.py

```diff
@@ -0,0 +1,225 @@
+from typing import Union, cast
+from abc import abstractmethod
+
+from ...types import TealType
+from ...errors import TealInputError
+from ..expr import Expr
+from ..seq import Seq
+from ..assert_ import Assert
+from ..substring import Suffix
+from ..int import Int
+from ..bytes import Bytes
+from ..unaryexpr import Itob
+from ..binaryexpr import GetByte, ExtractUint16, ExtractUint32, ExtractUint64
+from ..ternaryexpr import SetByte
+from .type import Type
+
+NUM_BITS_IN_BYTE = 8
+
+
+class Uint(Type):
+    def __init__(self, bit_size: int) -> None:
+        valueType = TealType.uint64 if bit_size <= 64 else TealType.bytes
+        super().__init__(valueType)
+        self.bit_size = bit_size
+
+    def has_same_type_as(self, other: Type) -> bool:
+        return isinstance(other, Uint) and self.bit_size == other.bit_size
+
+    def is_dynamic(self) -> bool:
+        return False
+
+    def bits(self) -> int:
+        return self.bit_size
+
+    def byte_length_static(self) -> int:
+        return self.bit_size // NUM_BITS_IN_BYTE
+
+    def get(self) -> Expr:
+        return self.stored_value.load()
+
+    @abstractmethod
+    def set(self, value: Union[int, Expr]) -> Expr:
+        pass
+
+    def __str__(self) -> str:
+        return "uint{}".format(self.bit_size)
+
+
+class Uint8(Uint):
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__(8)
+
+    def new_instance(self) -> "Uint8":
+        return Uint8()
+
+    def set(self, value: Union[int, Expr, "Uint8"]) -> Expr:
+        checked = False
+        if type(value) is int:
+            if value >= 2 ** 8:
+                raise TealInputError("Value exceeds Uint8 maximum: {}".format(value))
+            value = Int(value)
+            checked = True
+
+        if type(value) is Uint8:
+            value = value.get()
+            checked = True
+
+        if checked:
+            return self.stored_value.store(cast(Expr, value))
+
+        return Seq(
+            self.stored_value.store(cast(Expr, value)),
+            Assert(self.stored_value.load() < Int(2 ** self.bit_size)),
+        )
+
+    def decode(
+        self,
+        encoded: Expr,
+        *,
+        startIndex: Expr = None,
+        endIndex: Expr = None,
+        length: Expr = None
+    ) -> Expr:
+        if startIndex is None:
+            startIndex = Int(0)
+        return self.stored_value.store(GetByte(encoded, startIndex))
+
+    def encode(self) -> Expr:
+        return SetByte(Bytes(b"\x00"), Int(0), self.get())
+
+
+Uint8.__module__ = "pyteal"
+
+Byte = Uint8
```

Comment:
Good point. Originally I thought these two could use the same class, but since `uint8` and `byte` are different at the method signature level, I think `byte` needs to be its own class with its own `__str__`

Reply:
I can write a small pr to this one for `byte`, hope we can bring this out with more complete features. 
I am thinking of inherit directly from `uint8` and reuse most of the things, but I do not have a clear idea how to do conversion between `Byte` and `Uint8`.

Reply:
PR in #194